### PR TITLE
Hisat2-build: index construction w/ truly optional inputs 

### LINKED ISSUE #3648: Hisat2 build optional inputs
### Have you checked the docs?

- [X] [nf-core website: troubleshooting](https://nf-co.re/usage/troubleshooting)
- [X] [nf-core modules documentation](https://nf-co.re/docs/contributing/modules)

### Description of the bug

The Hisat2 build module decides whether to use information about splicing based on the available memory as higher memory is required to build splice aware indices. However, if this memory requirement is satisfied then splicesite and annotation files are made required arguments rather than optional. 

I suspect the optimal functionality would be for these arguments to be optional regardless of memory availability

### Command used and terminal output

_No response_

### Relevant files

_No response_

### System information

_No response_

### PR DESCRIPTION
**Pull request by JackCurragh**
## PR checklist

Closes #3648 

- [x] This comment contains a description of changes (with reason).

I have adapted the existing module to only use these optional inputs if they are provided. Previously the `--exons` and `--ss` parameters were optional dependent on available memory. Simpler indices were constructed if the memory requirement was not satisfied. This PR makes it so that nf-core users can intentionally build these 'non-splice aware' indices if they so choose (example use case is where there is no splicing).

### CODE REVIEW

**Review by jfy133 — CHANGES_REQUESTED** [review]
Please could you add more tests to check this is working?

**Comment by JackCurragh** [comment]
I actually completely missed that this was adding new functionality so didn't realise I needed tests. Getting the hang of things now...

**Comment by jfy133** [comment]
I guess it's not _new_ functionality, but if the tests had been written correctly in the first place the issue would've been picked up, so we do the good work here ;)

**Comment by JackCurragh** [comment]
Okay yes I get you. So now Linting is failing as the `test.yml` contains md5sums for empty files. These empty files are the output of `hisat2_extract_exons.py` on the[ sarscov2 gtf](https://github.com/nf-core/test-datasets/blob/modules/data/genomics/sarscov2/genome/genome.gtf) which does not contain any exons. So this is in fact the desired behaviour. 

This is failing for the one original test and then 2 of the 3 that I have added so I am not sure how this was passing the linting tests previously. Any idea how I should handle this? Change the test dataset? [For example.](https://github.com/nf-core/test-datasets/blob/modules/data/genomics/homo_sapiens/genome/genome.gtf)

**Review by jfy133 — COMMENTED** [review]
Could you 'clean up' the test.yml so it only has the files that come out of BUILD (i.e., not splicesites)?

And are your previous comments still valid now you've swwitched from SARS-COV to human?

**Review by jfy133 — COMMENTED** [review]
OK linting fail is API rate, but that's OK.

But I'm now concerned that `genome.splice_sites.txt` is not appearing in any of the test output not?


What's going on there... or was this module not working in the first place :eyes: 

We need someoneone from a pipeline using the module to help here I think...

@BarryDigby @drpatelh (but they might be on holiday, so might take a bit).

What you could do in the meantime for your own pipeline is to copy and paste the changes to `main.nf` inyour your pipeline's copy of the `main.nf` and run `nf-core modules patch hisat2/build` until this is solved.

**Comment by JackCurragh** [comment]
Oh, I removed that splicesite file from the test output because I thought that was what you were suggesting previously! Those files don't come from the build module itself. 

My understanding is that the intermediate processes (eg. HISAT2_EXTRACTSPLICESITES) should be tested in their own module tests so then just this modules outputs should be checked when testing these modules (eg. HISAT2_BUILD) 

I can add them back if you prefer @jfy133? 

p.s thank you for your patience :)

File: tests/modules/nf-core/hisat2/build_test/main.nf [hunk 1]
```diff
@@ -7,11 +7,33 @@ include { HISAT2_BUILD              } from '../../../../../modules/nf-core/hisat
...
 }
+
+workflow test_hisat2_build_fasta_only {
+    fasta = [ [id:'genome'],
+              file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
+            ]
+    gtf   = [ [id:'genome'],
+              file(params.test_data['homo_sapiens']['genome']['genome_gtf'], checkIfExists: true)
+            ]
+    HISAT2_EXTRACTSPLICESITES ( gtf )
```
Comment (jfy133) [inline tests/modules/nf-core/hisat2/build_test/main.nf:26]: Not needed to be executed here if not passed to HISAT2_BUILD in this test (and `gtf` could be removed I guess)

**Review by jfy133 — APPROVED** [review]
Ahh sorry, I guess that the original test.yml was bad then :sweat_smile:  (I guess I wasn't concentrating properly, my bad)! Yes my original request was still correct! I thin this looks OK to me now - Harshil or other RNA people can shout at me later if this is wrong ;)

**Comment by JackCurragh** [comment]
Okay, @jfy133, we are getting there! I have removed that unnecessary call from the tests, changes were approved and have all green ticks on actions. 

I think this means I can merge but just want to double check (and will know for next time)

File: tests/modules/nf-core/hisat2/build_test/main.nf [hunk 2]
```diff
@@ -7,11 +7,32 @@ include { HISAT2_BUILD              } from '../../../../../modules/nf-core/hisat
 
 workflow test_hisat2_build {
     fasta = [ [id:'genome'],
-              file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
+              file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
             ]
     gtf   = [ [id:'genome'],
-              file(params.test_data['sarscov2']['genome']['genome_gtf'], checkIfExists: true)
+              file(params.test_data['homo_sapiens']['genome']['genome_gtf'], checkIfExists: true)
```
Comment (maxulysse) [inline tests/modules/nf-core/hisat2/build_test/main.nf:13]: why changing from sarscov2 to homo_sapiens?

Comment (maxulysse) [reply tests/modules/nf-core/hisat2/build_test/main.nf:13]: ok, I read the comments, so makes sense

Comment (jfy133) [reply tests/modules/nf-core/hisat2/build_test/main.nf:13]: all OK then?

File: tests/modules/nf-core/hisat2/build_test/main.nf [hunk 3]
```diff
@@ -7,11 +7,32 @@ include { HISAT2_BUILD              } from '../../../../../modules/nf-core/hisat
...
 }
+
+workflow test_hisat2_build_fasta_only {
+    fasta = [ [id:'genome'],
+              file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
+            ]
+    gtf   = [ [id:'genome'],
+              file(params.test_data['homo_sapiens']['genome']['genome_gtf'], checkIfExists: true)
+            ]
+    HISAT2_BUILD ( fasta, [[],[]], [[],[]] )
```
Comment (JoseEspinosa) [inline tests/modules/nf-core/hisat2/build_test/main.nf:26]: Maybe change it to make it clearer that is the meta map?
```suggestion
    HISAT2_BUILD ( fasta, [[:],[]], [[:],[]] )
```

File: tests/modules/nf-core/hisat2/build_test/main.nf [hunk 4]
```diff
@@ -7,11 +7,32 @@ include { HISAT2_BUILD              } from '../../../../../modules/nf-core/hisat
...
+
+workflow test_hisat2_build_fasta_ss_only {
+    fasta = [ [id:'genome'],
+              file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
+            ]
+    gtf   = [ [id:'genome'],
+              file(params.test_data['homo_sapiens']['genome']['genome_gtf'], checkIfExists: true)
+            ]
+    HISAT2_EXTRACTSPLICESITES ( gtf )
+    HISAT2_BUILD ( fasta, [[],[]], HISAT2_EXTRACTSPLICESITES.out.txt )
```
Comment (JoseEspinosa) [inline tests/modules/nf-core/hisat2/build_test/main.nf:37]: Same here?
```suggestion
    HISAT2_BUILD ( fasta, [[:],[]], HISAT2_EXTRACTSPLICESITES.out.txt )
```

**Review by JoseEspinosa — COMMENTED** [review]
Maybe I am missing something but isn't the test for the `gtf` input missing?

**Comment by jfy133** [comment]
> Maybe I am missing something but isn't the test for the `gtf` input missing?

That's the first test no?

**Comment by JoseEspinosa** [comment]
> > Maybe I am missing something but isn't the test for the `gtf` input missing?
> 
> That's the first test no?

Yes, indeed. I think I need some holidays 😛

**Review by JoseEspinosa — APPROVED** [review]
LGTM! 🚀 
I just dropped a comment about the empty map declaration but this is more how I would do it but not really an issue 😄

**Comment by jfy133** [comment]
@JackCurragh you're GTG once making Jose's changes (assuming it doesn't break the testS)

**Comment by JoseEspinosa** [comment]
> @JackCurragh you're GTG once making Jose's changes (assuming it doesn't break the testS)

If you think it could break the test, then just go ahead

File: tests/modules/nf-core/hisat2/build_test/main.nf (hunk 2 above)
Comment (JackCurragh) [reply tests/modules/nf-core/hisat2/build_test/main.nf:13]: The sarscov2 annotation doesn't contain any exons to be extracted so the linting was complaining about having expected (intermediate) outputs that had empty md5sums. Indeed, this actually is a redundant change when I remove them from the `test.yml` as James advised. 

Should I revert back?

Comment (jfy133) [reply tests/modules/nf-core/hisat2/build_test/main.nf:13]: No, it's fine to switch! Doesn't amke a difference

Comment (JackCurragh) [reply tests/modules/nf-core/hisat2/build_test/main.nf:13]: Oh haha now I have switched back.

**Pull request merged**